llvm/lib/SYCLLowerIR/SYCLConditionalCallOnDevice.cpp

```diff
@@ -0,0 +1,108 @@
+//===-- SYCLConditionalCallOnDevice.cpp - SYCLConditionalCallOnDevice Pass
+//--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass performs transformations on functions which represent the conditional
+// call to application's callable object. The conditional call is based on the
+// SYCL device's aspects or architecture passed to the functions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLConditionalCallOnDevice.h"
+
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+
+using namespace llvm;
+
+PreservedAnalyses
+SYCLConditionalCallOnDevicePass::run(Module &M, ModuleAnalysisManager &) {
+  // find call_if_on_device_conditionally function
+  SmallVector<Function *, 4> FCallers;
+  for (Function &F : M.functions()) {
+    if (F.getName().contains("call_if_on_device_conditionally_helper") ||
```

Comment:
Can we use `startswith` instead of contains? Or prefix is also changing from overload to overload?

Reply:
The prefix doesn't change and stay always `_ZN4sycl3_V13ext6oneapi12experimental6detail` but I would prefer to not hardcode it

Reply:
Note that nothing prohibits a user from defining a function `call_if_on_device_conditionally_helper`. I understand that barcoding a prefix means we will have to change it every time header file changes, but in the current form we may react to user-defined functions and break a program.

Perhaps we want to annotate this function with an attribute (could be done through compile-time properties interface) to simplify its detection in the pass and make it more reliable.

Reply:
Oh, very good point, it really can be like that!